src/g_compatibility.c

```diff
@@ -0,0 +1,333 @@
+    "e26c1b6f4dfd90bb6533e6381bf61be5",
+    {{comp_vile, 1}, END}
+};
+
+static const comp_record_t arch_map01 = {
+    "1d37cbd32a1ecf4763437631e7b3c29a",
+    {{comp_vile, 1}, END}
+};
+
+static const comp_record_t ur_map06 = {
+    "cfb054683af1ed187d0565942d3dbb8f",
+    {{comp_vile, 1}, END}
+};
+
+static const comp_record_t *comp_database[] = {
+    &eternal_doom_25,
+    &hell_revealed_map19,
+    &roger_ritenour_phobos_map03,
+    &hell_revealed_map26,
+    &hell_to_pay_map14,
+    &hell_to_pay_map22,
+    &icarus_map24,
+    &plutonia2_map32,
+    &requiem_map23,
+    &the_waterfront_map01,
+    &gather2_map05_and_darkside_map01,
+    &reverie_map18,
+    &project_x_map14,
+    &archie_map01,
+    &seej_map01,
+    &sixpack2_map02,
+    &squadron_417_map21,
+    &mayhem_2013_map05,
+    &imps_are_ghost_gods_map01,
+    &confinement_map31,
+    &conf256b_map07,
+    &conf256b_map12,
+    &sunlust_map30,
+    &tnt_map30,
+    &intercep2_map03,
+    &skulltiverse_map02,
+    &tntr_map30,
+    &roomblow_e1m1,
+    &esp_map21,
+    &av_map07,
+    &sin2_9_map02,
+    &d2reload_map09,
+    &amoreupho_map02,
+    &dbp20_dnd_map07,
+    &arch_map01,
+    &ur_map06,
+};
+
+typedef struct
+{
+    byte bytes[16];
+    char string[33];
+} md5_checksum_t;
+
+static void MD5UpdateLump(int lump, struct MD5Context *md5)
+{
+    MD5Update(md5, W_CacheLumpNum(lump, PU_CACHE), W_LumpLength(lump));
+}
+
+static void GetLevelCheckSum(int lump, md5_checksum_t* cksum)
+{
+    struct MD5Context md5;
+
+    MD5Init(&md5);
+
+    MD5UpdateLump(lump + ML_LABEL, &md5);
+    MD5UpdateLump(lump + ML_THINGS, &md5);
+    MD5UpdateLump(lump + ML_LINEDEFS, &md5);
+    MD5UpdateLump(lump + ML_SIDEDEFS, &md5);
+    MD5UpdateLump(lump + ML_SECTORS, &md5);
+
+    // ML_BEHAVIOR when it becomes applicable to comp options
+
+    MD5Final(cksum->bytes, &md5);
+
+    for (int i = 0; i < sizeof(cksum->bytes); ++i)
+    {
+        sprintf(&cksum->string[i * 2], "%02x", cksum->bytes[i]);
+    }
+    cksum->string[32] = '\0';
+}
+
+// For casual players that aren't careful about setting complevels, this
+// function will apply comp options to automatically fix some issues that
+// appear when playing wads in mbf21 (since this is the default).
+
+void G_ApplyLevelCompatibility(int lump)
+{
+    if (demorecording || demoplayback || !mbf21)
+    {
+        return;
+    }
+
+    md5_checksum_t cksum;
+
+    GetLevelCheckSum(lump, &cksum);
+
+    I_Printf(VB_DEBUG, "Level checksum: %s", cksum.string);
+
+    for (int i = 0; i < arrlen(comp_database); ++i)
+    {
+        const comp_record_t *record = comp_database[i];
+
+        if (!strcasecmp(record->cksum_string, cksum.string))
+        {
+            for (int k = 0; record->options[k].comp != -1; ++k)
+            {
+                option_t option = record->options[k];
+
+                comp[option.comp] = option.value;
+
+                I_Printf(VB_INFO, "Automatically setting comp option %d %s.",
```

Comment:
`"Automatically setting comp option %d to %d.", option.comp, option.value)`

Reply:
Can we somehoe achieve that the settings are reset back when entering the next level? Do we want this, or is it a stupid idea?

Reply:
Done in https://github.com/fabiangreffrath/woof/pull/2015/commits/7b0766353f2d1ce263d304061d3e425faf2e98bf